From 6eac4615d18966dbc6b5796bafd9ad5777f336f3 Mon Sep 17 00:00:00 2001
From: Daniel Golle <daniel@makrotopia.org>
Date: Thu, 16 Mar 2023 16:31:51 +0530
Subject: [PATCH 049/281] OpenWrt:
 492-try-auto-mounting-ubi0-rootfs-in-init-do_mounts.c.patch

try auto-mounting ubi0:rootfs in init/do_mounts.c

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

Change-Id: I64d5e42242c75edc7b2446c25f008956a9f3f26d
Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>
---
 init/do_mounts.c | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/init/do_mounts.c b/init/do_mounts.c
index 811e94daf0a8..1378564e993e 100644
--- a/init/do_mounts.c
+++ b/init/do_mounts.c
@@ -446,7 +446,30 @@ void __init mount_block_root(char *name, int flags)
 out:
 	put_page(page);
 }
- 
+
+#ifdef CONFIG_MTD_ROOTFS_ROOT_DEV
+static int __init mount_ubi_rootfs(void)
+{
+	int flags = MS_SILENT;
+	int err, tried = 0;
+
+	while (tried < 2) {
+		err = do_mount_root("ubi0:rootfs", "ubifs", flags, \
+					root_mount_data);
+		switch (err) {
+			case -EACCES:
+				flags |= MS_RDONLY;
+				tried++;
+				break;
+			default:
+				return err;
+		}
+	}
+
+	return -EINVAL;
+}
+#endif
+
 #ifdef CONFIG_ROOT_NFS
 
 #define NFSROOT_TIMEOUT_MIN	5
@@ -578,6 +601,10 @@ void __init mount_root(void)
 			printk(KERN_ERR "VFS: Unable to mount root fs via SMB.\n");
 		return;
 	}
+#endif
+#ifdef CONFIG_MTD_ROOTFS_ROOT_DEV
+	if (!mount_ubi_rootfs())
+		return;
 #endif
 	if (ROOT_DEV == 0 && root_device_name && root_fs_names) {
 		if (mount_nodev_root() == 0)
-- 
2.17.1

